pkg/openfeature/event_executor.go

```diff
@@ -0,0 +1,381 @@
+
+	newProvider := providerReference{
+		featureProvider:   provider,
+		shutdownSemaphore: sem,
+	}
+
+	oldProvider := e.namedProviderReference[associatedClient]
+	e.namedProviderReference[associatedClient] = newProvider
+
+	return e.startListeningAndShutdownOld(newProvider, oldProvider)
+}
+
+// startListeningAndShutdownOld is a helper to start concurrent listening to new provider events and  invoke shutdown
+// hook of the old provider if it's not bound by another subscription
+func (e *eventExecutor) startListeningAndShutdownOld(newProvider providerReference, oldReference providerReference) error {
+
+	// check if this provider already actively handled - 1:N binding capability
+	if !isRunning(newProvider, e.activeSubscriptions) {
+		e.activeSubscriptions = append(e.activeSubscriptions, newProvider)
+
+		go func() {
+			v, ok := newProvider.featureProvider.(EventHandler)
+			if !ok {
+				return
+			}
+
+			// event handling of the new feature provider
+			for {
+				select {
+				case event := <-v.EventChannel():
+					e.eventChan <- eventPayload{
+						event:   event,
+						handler: newProvider.featureProvider,
+					}
+				case <-newProvider.shutdownSemaphore:
+					return
+				}
+			}
+		}()
+	}
+
+	// shutdown old provider handling
+
+	// check if this provider is still bound - 1:N binding capability
+	if isBound(oldReference, e.defaultProviderReference, maps.Values(e.namedProviderReference)) {
+		return nil
+	}
+
+	// drop from active references
+	for i, r := range e.activeSubscriptions {
+		if r == oldReference {
+			e.activeSubscriptions = append(e.activeSubscriptions[:i], e.activeSubscriptions[i+1:]...)
+		}
+	}
+
+	_, ok := oldReference.featureProvider.(EventHandler)
+	if !ok {
+		// no shutdown for non event handling provider
+		return nil
+	}
+
+	// avoid shutdown lockouts
+	select {
+	case oldReference.shutdownSemaphore <- "":
+		return nil
+	case <-time.After(200 * time.Millisecond):
+		return fmt.Errorf("old event handler %s timeout waiting for handler shutdown",
+			oldReference.featureProvider.Metadata().Name)
+	}
+}
+
+// startEventListener trigger the event listening of this executor
+func (e *eventExecutor) startEventListener() {
+	e.once.Do(func() {
+		go func() {
+			for payload := range e.eventChan {
+				e.triggerEvent(payload.event, payload.handler)
+			}
+		}()
+	})
+}
+
+// triggerEvent performs the actual event handling
+func (e *eventExecutor) triggerEvent(event Event, handler FeatureProvider) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	// first run API handlers
+	for _, c := range e.apiRegistry[event.EventType] {
+		e.executeHandler(*c, event)
+	}
+
+	// then run client handlers
+	var nameAssociations []string
+	for name, reference := range e.namedProviderReference {
+		if reference.featureProvider == handler {
+			nameAssociations = append(nameAssociations, name)
+		}
+	}
+
+	for _, nameAssociation := range nameAssociations {
+		for _, c := range e.scopedRegistry[nameAssociation].callbacks[event.EventType] {
```

Comment:
instead of going again through nameAssociations here, we could directly run the for at 314 in line 309

Reply:
Good suggestion. I went with this extract and execute it for easy reading. But yes, direct execution on the same loop is also fine.

Update with the latest commit :)